File: Data Structures & Algorithms/daily-temperatures/submission-0.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []

        res = [0] * len(temperatures) # 0 by default

        
        for i in range(len(temperatures)):
            # if we have an element on the top of the stack AND we see a value greater than it
            # update it with the index
            while stack and temperatures[stack[-1]] < temperatures[i]:
                index = stack.pop()
                res[index] = i - index
            stack.append(i)

        return res

File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        stack<int> st; // store index

        // iterate through the height 

        int res = 0;

        for (int i = 0; i < height.size(); ++i) {
            // monotonic stack
            // if there is an uptick in height, find the left bound
            while (!st.empty() && height[i] > height[st.top()]) {
                int bottom = st.top();
                st.pop();

                if (st.empty()) {
                    break;
                }

                int left = st.top();
                int width = i - left - 1;
                int boundedHeight = min(height[left], height[i]) - height[bottom];

                res += width * boundedHeight;
            }
            st.push(i);

        }
        return res;
    }
};
